File: P3/civitasruby/lib/civitas/casilla.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "tipo_casilla"
module Civitas
  class Casilla

    def initialize(nombre, titulo, cantidad, num_casilla_carcel, mazo, tipo)
      init
      @nombre = nombre
      @titulo = titulo
      @cantidad = cantidad
      @num_casilla_carcel = num_casilla_carcel
      @mazo = mazo
      @tipo = tipo
    end

    def self.new_nombre(nombre)
      Casilla.new(nombre, nil, -1, -1, nil, Tipo_casilla::DESCANSO)
    end

    def self.new_titulo_propiedad(titulo)
      Casilla.new(titulo.get_nombre, titulo, titulo.get_precio_compra, -1,nil, Tipo_casilla::CALLE)
    end

    def self.new_cantidad_nombre(cantidad, nombre)
      Casilla.new(nombre, nil, cantidad, -1, nil, nil)
    end

    def self.new_num_casilla_carcel_nombre(num_casilla_carcel, nombre)
      Casilla.new(nombre, nil, -1, num_casilla_carcel, nil, Tipo_casilla::JUEZ)
    end

    def self.new_mazo_nombre(mazo, nombre)
      Casilla.new(nombre, nil, -1, -1, mazo, Tipo_casilla::SORPRESA)
    end

    def init
      @importe = -1
      @nombre = ""
      @tipo = nil
      @mazo = nil
      @titulo_propiedad = nil
    end

    def recibe_jugador(iactual, todos)
      case tipo
      when CALLE
        recibeJugador_calle(iactual, todos) # hacer
      when IMPUESTO
        recibeJugador_impuesto(iactual, todos)
      when JUEZ
        recibeJugador_juez(iactual, todos)
      when SORPRESA
        recibeJugador_sorpresa(iactual, todos) # hacer
      else
        informe(iactual, todos)
      end
    end

    def recibeJugador_calle(iactual, todos)
      if (jugador_correcto(iactual, todos))
        informe(iactual, todos)
        jugador = todos.get(actual)
        if(!@titulo.tiene_propietario)
          jugador.puede_comprar_casilla
        else
          @titulo.tramitar_alquiler(jugador)
        end

      end
    end

    def recibeJugador_sorpresa(iactual, todos)
      if(jugador_correcto(iactual, todos))
        sorpresa = @mazo.siguiente
      end
      if(jugador_correcto(iactual, todos))
        informe(iactual, todos)
      end
      if(jugador_correcto(iactual, todos))
        sorpresa.aplicar_a_jugador(iactual, todos)
      end
    end

    def informe(actual, todos)
      if jugador_correcto(actual, todos)
        Diario.instance.ocurre_evento("El jugador #{todos.get(actual).get_nombre}
           ha caido en la casilla #{@nombre}. Informacion de la casilla: #{to_s}
          Informacion del jugador: #{todos.get(actual).to_s}")
      end
    end

    def jugador_correcto(actual, todos)
      actual >= 0 and actual < todos.length
    end

    @override
    def to_s
      info = "Nombre: #{@nombre}"
      if @tipo == Tipo_casilla::CALLE
        info += "\nPrecio: " + @importe
        propietario = @titulo_propiedad.get_propietario
        if propietario != nil
            info += "\nPropietario: " + propietario.get_nombre
        else
            info += "\nLa casilla no tiene dueño"
        end

      end

    end

    def recibeJugador_impuesto(actual, todos)
      if jugador_correcto(actual, todos)
        informe(actual, todos)
        todos.get(actual).paga_impuesto(importe)
      end
    end

    def recibeJugador_juez(actual, todos)
      if jugador_correcto(actual, todos)
        informe(actual, todos)
        todos.get(actual).encarcelar(@carcel)
      end
    end

    def get_nombre
      @nombre
    end

    def get_titulo
      @titulo_propiedad
    end

    private :informe, :init, :recibeJugador_calle, :recibeJugador_juez, :recibeJugador_sorpresa, :recibeJugador_impuesto

  end




end

File: P4/Civitas/src/civitas/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author Salva
 */
public class Casilla {

    //atributos de instancia
    private String nombre;

    //metodos
    Casilla(String nombre) {
        this.nombre = nombre;
    }

    protected void informe(int actual, ArrayList<Jugador> todos) {
        if (jugadorCorrecto(actual, todos)) {
            Diario.getInstance().ocurreEvento("El jugador " + todos.get(actual).getNombre() + " ha caido en la casilla " + nombre
                    + "\n=========\nInformación de la casilla:\n" + toString()
                    + "\n=========\nInformación del jugador:\n" + todos.get(actual).toString());

        }
    }

    public boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos) {
        return actual >= 0 && actual < todos.size();
    }

    @Override
    public String toString() {
        String info = "Nombre: " + nombre;
        return info;
    }

    public String getNombre() {
        return nombre;
    }
    
    void recibeJugador(int actual, ArrayList<Jugador> todos){
        informe(actual,todos);
    }
//    Casilla(TituloPropiedad titulo) {
//        init();
//        tituloPropiedad = titulo;
//        tipo = tipo.CALLE;
//        nombre = tituloPropiedad.getNombre();
//        importe = titulo.getPrecioCompra();
//    }
//
//    Casilla(float cantidad, String nombre) {
//        init();
//        this.nombre = nombre;
//        importe = cantidad;
//        tipo = TipoCasilla.IMPUESTO;
//    }
//
//    Casilla(int numCasillaCarcel, String nombre) {
//        init();
//        this.tipo = TipoCasilla.JUEZ;
//        this.nombre = nombre;
//        carcel = numCasillaCarcel;
//    }
//
//    Casilla(MazoSorpresas mazo, String nombre) {
//        init();
//        this.nombre = nombre;
//        this.mazo = mazo;
//        this.tipo = TipoCasilla.SORPRESA;
//    }

//    private void init() {
////        carcel = -1;
//        importe = -1;
//        nombre = "";
//        tipo = null;
//        mazo = null;
//        tituloPropiedad = null;
//
//    }

//    private void recibeJugador_impuesto(int actual, ArrayList<Jugador> todos) {
//        if (jugadorCorrecto(actual, todos)) {
//            informe(actual, todos);
//            todos.get(actual).pagaImpuesto(importe);
//        }
//    }
//
//    private void recibeJugador_juez(int actual, ArrayList<Jugador> todos) {
//        if (jugadorCorrecto(actual, todos)) {
//            informe(actual, todos);
//            todos.get(actual).encarcelar(carcel);
//        }
//    }


//    TituloPropiedad getTitulo() {
//        return tituloPropiedad;
//    }
//
//    void recibeJugador(int actual, ArrayList<Jugador> todos) {
//        switch (tipo) {
//            case CALLE:
//                this.recibeJugador_calle(actual, todos);
//                break;
//            case IMPUESTO:
//                this.recibeJugador_impuesto(actual, todos);
//                break;
//            case JUEZ:
//                this.recibeJugador_juez(actual, todos);
//                break;
//            case SORPRESA:
//                this.recibeJugador_sorpresa(actual, todos);
//                break;
//            default:
//                informe(actual, todos);
//                break;
//        }
//    }
//
//    private void recibeJugador_calle(int actual, ArrayList<Jugador> todos) {
//        if (jugadorCorrecto(actual, todos)) {
//            informe(actual, todos);
//            Jugador jugador = todos.get(actual);
//            if (!tituloPropiedad.tienePropietario()) {
//                jugador.puedeComprarCasilla();
//            } else {
//                tituloPropiedad.tramitarAlquiler(jugador);
//            }
//        }
//
//    }
//
//    private void recibeJugador_sorpresa(int actual, ArrayList<Jugador> todos) {
//        if (jugadorCorrecto(actual, todos)) {
//            Sorpresa sorpresa = mazo.siguiente();
//            informe(actual, todos);
//            sorpresa.aplicarAJugador(actual, todos);
//        }
//    }
}
